src/certAnalyser/analyserPool.go

```diff
@@ -186,69 +198,135 @@ func worker(msgs <-chan amqp.Delivery, es *elastigo.Conn) {
+		// to end up here either there was no leaf certificate retrieved
+		// or it was retrieved but it was not valid so we must check the remainder of the chain
+		for i, cert := range intermediates {
 
-		if c.IsCA {
-			dnsName = c.Subject.CommonName
-		}
+			inter := append(intermediates[:i], intermediates[i+1:]...)
 
-		opts := x509.VerifyOptions{
-			DNSName:       dnsName,
-			Intermediates: inter,
-			//will add rootCAs from cfg file
+			HandleCertChain(cert, inter, &curTS, chain.Domain, chain.IP)
+			//should we break if/when this validates?
 		}
 
-		var chains [][]*x509.Certificate
+	}
+}
 
-		chains, err := c.Verify(opts)
+func HandleCertChain(certificate *x509.Certificate, intermediates []*x509.Certificate, curTS *TrustStore, domain, IP string) bool {
 
-		if err == nil {
-			for i, cert := range chains[0] {
+	valInfo := &certValidationInfo{}
+
+	valInfo.IsValid = true
+
+	inter := x509.NewCertPool()
+	for _, in := range intermediates {
+		inter.AddCert(in)
+	}
+
+	dnsName := domain
+
+	if certificate.IsCA {
+		dnsName = certificate.Subject.CommonName
+	}
+
+	opts := x509.VerifyOptions{
+		DNSName:       dnsName,
+		Intermediates: inter,
+		Roots:         curTS.Certs,
+	}
+
+	chains, err := certificate.Verify(opts)
+
+	if err == nil {
+
+		for i, ch := range chains {
+			for _, cert := range ch {
+
+				log.Println("Trust chain no:" + strconv.Itoa(i))
 
 				parentSignature := ""
-				if cert.Issuer.CommonName != "" && len(certs) > i+1 {
-					parentSignature = SHA256Hash(certs[i+1].Raw)
+				c := getFirstParent(cert, ch)
+
+				if c != nil {
+					parentSignature = SHA256Hash(c.Raw)
+				} else {
+					log.Println("could not retrieve parent for " + dnsName)
 				}
-				pushCertificate(cert, parentSignature, chain.Domain, chain.IP, "", es)
+				pushCertificate(cert, parentSignature, domain, IP, curTS.Name, valInfo)
 			}
-			break
+		}
+		return true
+	} else {
+		if len(chains) > 0 {
+			log.Println("validation error but validation chain populated for: " + dnsName)
+		}
+
+		valInfo.ValidationError = err.Error()
+		valInfo.IsValid = false
+
+		parentSignature := ""
+		c := getFirstParent(certificate, intermediates)
+
+		if c != nil {
+			parentSignature = SHA256Hash(c.Raw)
 		} else {
-			parentSignature := ""
-			if c.Issuer.CommonName != "" && len(certs) > i+1 {
-				parentSignature = SHA256Hash(certs[i+1].Raw)
-			}
-			pushCertificate(c, parentSignature, chain.Domain, chain.IP, err.Error(), es)
+			log.Println("could not retrieve parent for " + dnsName)
+		}
+
+		pushCertificate(certificate, parentSignature, domain, IP, curTS.Name, valInfo)
+
+		return false
+	}
+
+}
+
+//Returns the first parent found for a certificate in a given certificate list ( does not verify signature)
+func getFirstParent(cert *x509.Certificate, certs []*x509.Certificate) *x509.Certificate {
+	for _, c := range certs {
+		if cert.Issuer.CommonName == c.Subject.CommonName { //TODO : consider changing this check with validating check
```

Comment:
Does this need to verify signatures, considering the chain is already validated by Go?